File: 3310-remove-methods-from-project/3310-remove-methods-from-project.java
class Solution {
    private void dfs(List<List<Integer>> adj, boolean[] vis, int node) {
        if (vis[node])
            return;
        
        vis[node] = true;
        for (int nbr : adj.get(node))
            dfs(adj, vis, nbr);
    }

    public List<Integer> remainingMethods(int n, int k, int[][] invocations) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i=0; i<n; i++)
            adj.add(new ArrayList<>());
        
        for (int[] inv : invocations)
            adj.get(inv[0]).add(inv[1]);

        boolean[] vis = new boolean[n];
        dfs(adj, vis, k);

        Queue<Integer> q = new LinkedList<>();
        for (int i=0; i<n; i++) {
            if (!vis[i]) {
                q.offer(i);
            }
        }

        Set<Integer> set = new HashSet<>();
        boolean flag = false;
        while (!q.isEmpty()) {
            int node = q.poll();
            
            set.add(node);

            if (vis[node])
                flag = true;

            for (int nbr : adj.get(node)) {
                if (!set.contains(nbr)) {
                    q.add(nbr);
                }
            }
        } 

        if (flag)
            for (int i=0; i<n; i++)
                if (vis[i]) 
                    set.add(i);

        return new ArrayList<>(set);
    }
}

File: 3310-remove-methods-from-project/3310-remove-methods-from-project.rs
use std::collections::{HashSet, VecDeque};

impl Solution {
    fn dfs(adj: &Vec<Vec<usize>>, vis: &mut Vec<bool>, node: usize) {
        if vis[node] {
            return;
        }

        vis[node] = true;

        for &nbr in &adj[node] {
            Self::dfs(adj, vis, nbr);
        }
    }

    pub fn remaining_methods(
        n: i32,
        k: i32,
        invocations: Vec<Vec<i32>>,
    ) -> Vec<i32> {
        let n = n as usize;
        let k = k as usize;

        let mut adj = vec![Vec::new(); n];

        for inv in invocations {
            adj[inv[0] as usize].push(inv[1] as usize);
        }

        let mut vis = vec![false; n];
        Self::dfs(&adj, &mut vis, k);

        let mut q = VecDeque::new();
        for i in 0..n {
            if !vis[i] {
                q.push_back(i);
            }
        }

        let mut set = HashSet::new();
        let mut flag = false;

        while let Some(node) = q.pop_front() {
            if !set.insert(node) {
                continue;
            }

            if vis[node] {
                flag = true;
            }

            for &nbr in &adj[node] {
                if !set.contains(&nbr) {
                    q.push_back(nbr);
                }
            }
        }

        if flag {
            for i in 0..n {
                if vis[i] {
                    set.insert(i);
                }
            }
        }

        set.into_iter().map(|x| x as i32).collect()
    }
}